src/librustc_typeck/check/mod.rs

```diff
@@ -4364,11 +4364,17 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
         let lifetime_defs = segment.map_or(&[][..], |(_, generics)| &generics.regions);
         if lifetimes.len() > lifetime_defs.len() {
             let span = lifetimes[lifetime_defs.len()].span;
-            span_err!(self.tcx.sess, span, E0088,
-                      "too many lifetime parameters provided: \
-                       expected {}, found {}",
-                      count(lifetime_defs.len()),
-                      count(lifetimes.len()));
+            let mut err = struct_span_err!(self.tcx.sess, span, E0088,
+                                           "too many lifetime parameters provided: \
+                                            expected {}, found {}",
+                                           count(lifetime_defs.len()),
+                                           count(lifetimes.len()));
+            match lifetime_defs.len() {
+                0 => err.span_label(span, &format!("unexpected lifetime parameter")),
+                _ => err.span_label(span,  &format!("{} lifetime parameters expected",
```

Comment:
Now that we handle multiple cases, can you add to the unit test to check for these new cases?